From f554b06ca88ba1f1ebfc547a7e5db1067d36efc8 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Wed, 23 Oct 2024 17:37:20 +0200
Subject: [PATCH 1321/1695] drivers: stm32_hpdma: fix and factorize semaphore
 handling for HPDMA

Fix the semaphore handling by properly releasing non-secure accessible
semaphores and factorize semaphore handling in
handle_available_semaphores().

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 2aab2870c2495cc5f9bad23668e52746e14b0a8e drivers: stm32_hpdma: fix and factorize semaphore handling for HPDMA]
Change-Id: Idf43b39be8be6e5fe19c6be84f094cf01d10baae
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/413675
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/drivers/stm32_hpdma.c | 114 ++++++++++++++++++++-----------------
 1 file changed, 62 insertions(+), 52 deletions(-)

diff --git a/core/drivers/stm32_hpdma.c b/core/drivers/stm32_hpdma.c
index 67101fd3c..86b8aaf4a 100644
--- a/core/drivers/stm32_hpdma.c
+++ b/core/drivers/stm32_hpdma.c
@@ -120,43 +120,74 @@ static TEE_Result stm32_hpdma_syscfg_set_arcr(struct hpdma_pdata *hpdma_d)
 	return TEE_SUCCESS;
 }
 
-static TEE_Result apply_rif_config(struct hpdma_pdata *hpdma_d, bool is_tdcid)
+static TEE_Result handle_available_semaphores(struct hpdma_pdata *hpdma_d)
 {
-	TEE_Result res = TEE_ERROR_ACCESS_DENIED;
+	TEE_Result res = TEE_ERROR_GENERIC;
 	uint32_t cidcfgr = 0;
 	unsigned int i = 0;
 
-	if (!hpdma_d->conf_data)
-		return TEE_SUCCESS;
-
-	for (i = 0; i < HPDMA_RIF_CHANNELS; i++) {
+	for (i = 0 ; i < HPDMA_RIF_CHANNELS; i++) {
 		if (!(BIT(i) & hpdma_d->conf_data->access_mask[0]))
 			continue;
-		/*
-		 * When TDCID, OP-TEE should be the one to set the CID filtering
-		 * configuration. Clearing previous configuration prevents
-		 * undesired events during the only legitimate configuration.
-		 */
-		if (is_tdcid)
-			io_clrbits32(hpdma_d->base + _HPDMA_CIDCFGR(i),
-				     _HPDMA_CIDCFGR_CONF_MASK);
 
 		cidcfgr = io_read32(hpdma_d->base + _HPDMA_CIDCFGR(i));
 
-		/* Check if the channel is in semaphore mode */
 		if (!stm32_rif_semaphore_enabled_and_ok(cidcfgr, RIF_CID1))
 			continue;
 
-		/* If not TDCID, we want to acquire semaphores assigned to us */
-		res = stm32_rif_acquire_semaphore(hpdma_d->base +
-						  _HPDMA_SEMCR(i),
-						  HPDMA_NB_MAX_CID_SUPPORTED);
-		if (res) {
-			EMSG("Couldn't acquire semaphore for channel %u", i);
-			clk_disable(hpdma_d->hpdma_clock);
-			return res;
+		if (!(io_read32(hpdma_d->base + _HPDMA_SECCFGR) & BIT(i))) {
+			res =
+			stm32_rif_release_semaphore(hpdma_d->base +
+						    _HPDMA_SEMCR(i),
+						    HPDMA_NB_MAX_CID_SUPPORTED);
+			if (res) {
+				EMSG("Cannot release semaphore for resource %"PRIu32,
+				     i);
+				return res;
+			}
+		} else {
+			res =
+			stm32_rif_acquire_semaphore(hpdma_d->base +
+						    _HPDMA_SEMCR(i),
+						    HPDMA_NB_MAX_CID_SUPPORTED);
+			if (res) {
+				EMSG("Cannot acquire semaphore for resource %"PRIu32,
+				     i);
+				return res;
+			}
 		}
 	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result apply_rif_config(struct hpdma_pdata *hpdma_d, bool is_tdcid)
+{
+	TEE_Result res = TEE_ERROR_ACCESS_DENIED;
+	unsigned int i = 0;
+
+	if (!hpdma_d->conf_data)
+		return TEE_SUCCESS;
+
+	if (is_tdcid) {
+		for (i = 0; i < HPDMA_RIF_CHANNELS; i++) {
+			if (!(BIT(i) & hpdma_d->conf_data->access_mask[0]))
+				continue;
+			/*
+			 * When TDCID, OP-TEE should be the one to set the CID
+			 * filtering configuration. Clearing previous
+			 * configuration prevents undesired events during the
+			 * only legitimate configuration.
+			 */
+			io_clrbits32(hpdma_d->base + _HPDMA_CIDCFGR(i),
+				     _HPDMA_CIDCFGR_CONF_MASK);
+		}
+	} else {
+		res = handle_available_semaphores(hpdma_d);
+		if (res)
+			panic();
+	}
+
 	/* Security and privilege RIF configuration */
 	io_clrsetbits32(hpdma_d->base + _HPDMA_PRIVCFGR, _HPDMA_PRIVCFGR_MASK &
 			hpdma_d->conf_data->access_mask[0],
@@ -166,8 +197,10 @@ static TEE_Result apply_rif_config(struct hpdma_pdata *hpdma_d, bool is_tdcid)
 			hpdma_d->conf_data->sec_conf[0]);
 
 	/* Skip CID/semaphore configuration if not in TDCID state. */
-	if (!is_tdcid)
+	if (!is_tdcid) {
+		res = TEE_SUCCESS;
 		goto end;
+	}
 
 	for (i = 0; i < HPDMA_RIF_CHANNELS; i++) {
 		uint32_t cid_conf = hpdma_d->conf_data->cid_confs[i];
@@ -194,33 +227,6 @@ static TEE_Result apply_rif_config(struct hpdma_pdata *hpdma_d, bool is_tdcid)
 
 		io_clrsetbits32(hpdma_d->base + _HPDMA_CIDCFGR(i),
 				_HPDMA_CIDCFGR_CONF_MASK, cid_conf);
-
-		cidcfgr = io_read32(hpdma_d->base + _HPDMA_CIDCFGR(i));
-
-		/*
-		 * Take semaphore if the resource is in semaphore
-		 * mode and secured.
-		 */
-		if (!stm32_rif_semaphore_enabled_and_ok(cidcfgr, RIF_CID1) ||
-		    !(io_read32(hpdma_d->base + _HPDMA_SECCFGR) & BIT(i))) {
-			res =
-			stm32_rif_release_semaphore(hpdma_d->base +
-						    _HPDMA_SEMCR(i),
-						    HPDMA_NB_MAX_CID_SUPPORTED);
-			if (res) {
-				EMSG("Couldn't release semaphore for res%u", i);
-				return TEE_ERROR_ACCESS_DENIED;
-			}
-		} else {
-			res =
-			stm32_rif_acquire_semaphore(hpdma_d->base +
-						    _HPDMA_SEMCR(i),
-						    HPDMA_NB_MAX_CID_SUPPORTED);
-			if (res) {
-				EMSG("Couldn't acquire semaphore for res%u", i);
-				return TEE_ERROR_ACCESS_DENIED;
-			}
-		}
 	}
 
 	/*
@@ -230,6 +236,10 @@ static TEE_Result apply_rif_config(struct hpdma_pdata *hpdma_d, bool is_tdcid)
 	io_clrsetbits32(hpdma_d->base + _HPDMA_RCFGLOCKR, _HPDMA_RCFGLOCKR_MASK,
 			hpdma_d->conf_data->lock_conf[0]);
 
+	res = handle_available_semaphores(hpdma_d);
+	if (res)
+		panic();
+
 end:
 	if (IS_ENABLED(CFG_TEE_CORE_DEBUG)) {
 		/* Check that RIF config are applied, panic otherwise */
@@ -248,7 +258,7 @@ end:
 		}
 	}
 
-	return TEE_SUCCESS;
+	return res;
 }
 
 static TEE_Result parse_dt(const void *fdt, int node,
-- 
2.47.2

